### PR TITLE
Block cloning tests.

### PR DESCRIPTION
**Pull request by pjd**
Mostly focusing on testing various corner cases.

### Motivation and Context
There are a lot of corner cases related to block cloning that have to be handled in the kernel.
Those additional tests try to exercise as many of those corner cases as possible. 

### Description
The new tests take a long time to execute (~5h), so they can be executed through a dedicated runfile - bclone.run.
In common.run runfile there are all the tests, except for the corner case tests - for those only 100 randomly selected tests will be executed.

### How Has This Been Tested?
The tests pass on FreeBSD and Linux. 

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)
- [x] Tests

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by robn** [comment]
This is great @pjd, thanks. If no one else gets in I'll try to take a good look through over the weekend.

> The new tests take a long time to execute (~5h), so they can be executed through a dedicated runfile - bclone.run.

Is there a useful subset that we could put on the common or sanity lists?

File: tests/runfiles/common.run [hunk 1]
```diff
@@ -53,6 +53,24 @@ tags = ['functional', 'arc']
...
+    'bclone_diffprops_all',
+    'bclone_diffprops_checksum',
+    'bclone_diffprops_compress',
+    'bclone_diffprops_copies',
+    'bclone_diffprops_recordsize',
+    'bclone_prop_sync',
+    'bclone_samefs_corner_cases_limited',
+    'bclone_samefs_data',
+    'bclone_samefs_embedded',
+    'bclone_samefs_hole']
```
Comment (behlendorf) [inline tests/runfiles/common.run:70]: It looks like a bunch of these new tests need to be added to PR commits.

There are also a bunch of existing tests in the `linux.run` file.  If possible it'd be good to adapt these so they can be run on FreeBSD as well.

```
[tests/functional/block_cloning:Linux]
tests = ['block_cloning_copyfilerange', 'block_cloning_copyfilerange_partial',
    'block_cloning_copyfilerange_fallback',
    'block_cloning_ficlone', 'block_cloning_ficlonerange',
    'block_cloning_ficlonerange_partial',
    'block_cloning_disabled_copyfilerange', 'block_cloning_disabled_ficlone',
    'block_cloning_disabled_ficlonerange',
    'block_cloning_copyfilerange_cross_dataset',
    'block_cloning_copyfilerange_fallback_same_txg']
tags = ['functional', 'block_cloning']
```

**Review by behlendorf — COMMENTED** [review]
@pjd it's great to see these additional comprehensive tests.  If you'd like to test them independently in the CI you could additionally create a `bclone.run` file and then tweak the top level TESTS files like this.  That would let us get them running, see how long they take in the CI environment, and maybe let us pick a subset to run regularly.

```diff
-#TEST_ZFSTESTS_RUNFILE="linux.run"
+TEST_ZFSTESTS_RUNFILE="bclone.run"
```

File: tests/zfs-tests/tests/functional/bclone/TODO [hunk 2]
```diff
@@ -0,0 +1,6 @@
+- If dedup enabled, block_cloning uses dedup.
+- block_cloning works only with the same encryption key
+- check when block cloning doesn't suppose to work
+- check block cloning between two different pools
+- test with sync=always,default,disabled
+- block cloning from a snapshot
```
Comment (oromenahar) [inline tests/zfs-tests/tests/functional/bclone/TODO:6]: nice work, maybe have time at the weekend as well.

don't know but is it important to rebase this, because of some new commit in the master? [diff](https://github.com/pjd/openzfs/compare/brt-tests...openzfs:zfs:master) there is across encrypted devices new for example

Comment (oromenahar) [reply tests/zfs-tests/tests/functional/bclone/TODO:6]: Also some tests are added for encrypted datasets.

File: TEST [hunk 3]
```diff
@@ -34,7 +34,7 @@
 #TEST_ZFSTESTS_DISKSIZE="8G"
 #TEST_ZFSTESTS_ITERS="1"
 #TEST_ZFSTESTS_OPTIONS="-vx"
-#TEST_ZFSTESTS_RUNFILE="linux.run"
+TEST_ZFSTESTS_RUNFILE="bclone.run"
 #TEST_ZFSTESTS_TAGS="functional"
```
Comment (behlendorf) [inline TEST:38]: The CI invoked the test suite like so.  It looks like nothing ran because there is no "functional" tag for the bclone group in the new `bclone.run` file.  I'd suggest adding the tag and updating the PR to give it another try.

```
zfs-tests.sh -vxR -r bclone.run -d /mnt/ -I 1 -T functional
```

**Comment by pjd** [comment]
> This is great @pjd, thanks. If no one else gets in I'll try to take a good look through over the weekend.
> 
> > The new tests take a long time to execute (~5h), so they can be executed through a dedicated runfile - bclone.run.
> 
> Is there a useful subset that we could put on the common or sanity lists?

Yes, I've added all short tests to common.run plus 100 randomly selected at each run corner case tests.

File: tests/zfs-tests/include/libtest.shlib [hunk 4]
```diff
@@ -2629,6 +2629,38 @@ function get_config
...
+	for (( i = 0; i < $count; i++ )); do
+		while true; do
+			((value = RANDOM % (max + 1)))
+			if ((value >= min)); then
+				break
+			fi
+		done
+		echo $value
+	done
+}
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib:2662]: There an alternate version of this function already in `zfs-tests/include/math.shlib` called `random_int_between`.  Let's unify them and extend the existing version to take that third "How Many" option.

```sh

# A simple function to get a random number between two bounds (inclusive)
#
# Note since we're using $RANDOM, $min+32767 is the largest number we
# can accept as the upper bound.
#
# $1 lower bound
# $2 upper bound
function random_int_between
{
        typeset -i min=$1
        typeset -i max=$2

        echo $(( (RANDOM % (max - min + 1)) + min ))
}
```

File: tests/zfs-tests/tests/functional/bclone/TODO [hunk 5]
```diff
@@ -0,0 +1,5 @@
+- If dedup enabled, block_cloning uses dedup.
+- block_cloning works only with the same encryption key
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/bclone/TODO:2]: Perhaps this one can be removed, `block_cloning_cross_enc_dataset.ksh` was added for this.

File: tests/zfs-tests/tests/functional/bclone/bclone_common.kshlib [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2023 by Pawel Jakub Dawidek
+#
+
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/bclone/bclone_common.kshlib:25]: A generic `verify_block_cloning` check would be nice to have.  You'll need to call this in all of the new tests.

```ksh
function verify_block_cloning
{
        if is_linux && [[ $(linux_version) -lt $(linux_version "4.5") ]]; then
                log_unsupported "copy_file_range not available before Linux 4.5"
        fi
}
```

You'll also need to add all the new tests to this block in `./test-runner/bin/zts-report.py.in`

```py
elif sys.platform.startswith('linux'):
    maybe.update({
```

File: tests/zfs-tests/tests/functional/bclone/bclone_common.kshlib [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    typeset -r original="${srcdir}/original"
+    typeset -r clone="${dstdir}/clone"
+
+    log_note "Testing file copy with datatype $datatype, file size $filesize, embedded $embedded"
+
+    # Create a test file with known content.
+    case $datatype in
+        random|text)
+            log_must zpool sync $TESTPOOL
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/bclone/bclone_common.kshlib:116]:
```suggestion
            sync_pool $TESTPOOL
```

Here and elsewhere.  We added a wrapper for this a while back and have been trying to use it consistently.

File: tests/test-runner/bin/zts-report.py.in [hunk 8]
```diff
@@ -263,13 +263,50 @@ if sys.platform.startswith('freebsd'):
...
+        'block_cloning/block_cloning_cross_enc_dataset':
+            ['SKIP', cfr_cross_reason],
+        'block_cloning/block_cloning_disabled_copyfilerange':
+            ['SKIP', cfr_reason],
+        'block_cloning/block_cloning_lwb_buffer_overflow':
+            ['SKIP', cfr_reason],
+        'block_cloning/block_cloning_replay':
+            ['SKIP', cfr_reason],
+        'block_cloning/block_cloning_replay_encrypted':
+            ['SKIP', cfr_reason],
```
Comment (behlendorf) [inline tests/test-runner/bin/zts-report.py.in:309]: I believe you need to the `.ksh` extension here.  These aren't being recognized as SKIPS.

```
Tests with results other than PASS that are unexpected:
    SKIP bclone/bclone_crossfs_corner_cases_limited (expected PASS)
    SKIP bclone/bclone_crossfs_data (expected PASS)
    SKIP bclone/bclone_crossfs_embedded (expected PASS)
    SKIP bclone/bclone_crossfs_hole (expected PASS)
    SKIP bclone/bclone_diffprops_all (expected PASS)
    SKIP bclone/bclone_diffprops_checksum (expected PASS)
    SKIP bclone/bclone_diffprops_compress (expected PASS)
    SKIP bclone/bclone_diffprops_copies (expected PASS)
    SKIP bclone/bclone_diffprops_recordsize (expected PASS)
    SKIP bclone/bclone_prop_sync (expected PASS)
    SKIP bclone/bclone_samefs_corner_cases_limited (expected PASS)
    SKIP bclone/bclone_samefs_data (expected PASS)
    SKIP bclone/bclone_samefs_embedded (expected PASS)
    SKIP bclone/bclone_samefs_hole (expected PASS)
```

The style checker also has [some formatting complaints]((https://github.com/openzfs/zfs/actions/runs/7294586309/job/19879696271?pr=15631) to address.  Then I _think_ this should be ready to go.

**Review by behlendorf — APPROVED** [review]
